kvbc/src/v4blockchain/v4_blockchain.cpp

```diff
@@ -142,4 +144,123 @@ void KeyValueBlockchain::addGenesisBlockKey(categorization::Updates& updates) co
           concordUtils::toBigEndianStringBuffer(block_chain_.getGenesisBlockId()), stale_on_update});
 }
+
+template <typename UPDATESTYPE, typename VALUETYPE>
+std::optional<VALUETYPE> KeyValueBlockchain::getValueFromUpdate(BlockId block_id,
+                                                                const std::string &key,
+                                                                const UPDATESTYPE &updates) const {
+  const auto valit = updates.kv.find(key);
+  if (valit == updates.kv.cend()) {
+    return std::nullopt;
+  }
+  auto v = categorization::DbValue{};
+  deserialize(valit->second, v);
+  if (v.deleted) {
+    return std::nullopt;
+  }
+  return VALUETYPE{{block_id, std::move(v.data)}};
+}
+std::optional<categorization::Value> KeyValueBlockchain::getValueFromUpdate(
+    BlockId block_id, const std::string &key, const categorization::BlockMerkleInput &updates) const {
+  return getValueFromUpdate<categorization::BlockMerkleInput, categorization::MerkleValue>(block_id, key, updates);
+}
+std::optional<categorization::Value> KeyValueBlockchain::getValueFromUpdate(
+    BlockId block_id, const std::string &key, const categorization::VersionedInput &updates) const {
+  return getValueFromUpdate<categorization::VersionedInput, categorization::VersionedValue>(block_id, key, updates);
+}
+std::optional<categorization::Value> KeyValueBlockchain::getValueFromUpdate(
+    BlockId block_id, const std::string &key, const categorization::ImmutableInput &updates) const {
+  return getValueFromUpdate<categorization::ImmutableInput, categorization::ImmutableValue>(block_id, key, updates);
+}
+
+// BLOCKS_CF
+std::optional<categorization::Value> KeyValueBlockchain::get(const std::string &category_id,
+                                                             const std::string &key,
+                                                             BlockId block_id) const {
+  auto updates_in_block = block_chain_.getBlockUpdates(block_id);
+  if (!updates_in_block) {
+    return std::nullopt;
+  }
+  const auto &kv_updates = updates_in_block->categoryUpdates(category_id);
+  if (!kv_updates) {
+    return std::nullopt;
+  }
+  std::optional<categorization::Value> ret;
+  std::visit(
+      [this, block_id, &key, &ret](const auto &specific_cat_updates) {
+        ret = this->getValueFromUpdate(block_id, key, specific_cat_updates);
+      },
+      (*kv_updates).get());
+  return ret;
+}
+
+// LATEST_KEYS_CF
+std::optional<categorization::Value> KeyValueBlockchain::getLatest(const std::string &category_id,
+                                                                   const std::string &key) const {
+  BlockId latest_block_id = block_chain_.getLastReachable();
+  return latest_keys_.getValue(latest_block_id, category_id, detail::Blockchain::generateKey(latest_block_id), key);
+}
+
+// BLOCKS_CF
+void KeyValueBlockchain::multiGet(const std::string &category_id,
+                                  const std::vector<std::string> &keys,
```

Comment:
Don't you want to implement this method as you did with the multigetLatest, which uses the getLatest in the loop?
In practice, I don't think this method is ever called by a real client.

Reply:
I meant to just call ```KeyValueBlockchain::get``` in a loop

Reply:
For BLOCK_CF as there is no timestamp, I used the multiGet of native_client. This will make sure that rocksDB runs the query and we avoid multiple DB calls. This will help a lot when DB is in a separate box.

Reply:
i am keeping it for now as we discussed.